jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableUpcallHandler.java

```diff
@@ -0,0 +1,180 @@
+/*[INCLUDE-IF JAVA_SPEC_VERSION >= 16]*/
+/*******************************************************************************
+ * Copyright (c) 2021, 2022 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+package jdk.internal.foreign.abi;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodType;
+
+import jdk.incubator.foreign.FunctionDescriptor;
+import jdk.incubator.foreign.MemoryLayout;
+/*[IF JAVA_SPEC_VERSION >= 18]*/
+import jdk.incubator.foreign.NativeSymbol;
+/*[ENDIF] JAVA_SPEC_VERSION >= 18 */
+import jdk.incubator.foreign.ResourceScope;
+
+/**
+ * The counterpart in OpenJDK is replaced with this class that wrap up
+ * an upcall handler enabling the native call to the java code at runtime.
+ */
+/*[IF JAVA_SPEC_VERSION >= 18]*/
+public class ProgrammableUpcallHandler { // UpcallHandler is removed in Java 18
+/*[ELSE] JAVA_SPEC_VERSION >= 18 */
+public class ProgrammableUpcallHandler implements UpcallHandler {
+/*[ENDIF] JAVA_SPEC_VERSION >= 18 */
+
+	private MemoryLayout[] argLayoutArray;
```

Comment:
can argLayoutArray and realReturnLayout be final as well?

Reply:
Agreed and fixed with `final` as suggested.